lib/src/front_end/ast_node_visitor.dart

```diff
@@ -425,37 +427,100 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
 
   @override
   void visitForStatement(ForStatement node) {
-    throw UnimplementedError();
+    token(node.forKeyword);
+    writer.split();
+    var forKeyword = writer.pop();
+
+    // Treat the for loop parts sort of as an argument list where each clause
+    // is a separate argument. This means that when they split, they split like:
+    //
+    // ```
+    // for (
+    //   initializerClause;
+    //   conditionClause;
+    //   incrementClause
+    // ) {
+    //   body;
+    // }
+    // ```
+    var partsList =
+        DelimitedListBuilder(this, const ListStyle(commas: Commas.none));
+    partsList.leftBracket(node.leftParenthesis);
+
+    var forLoopParts = node.forLoopParts;
+    switch (forLoopParts) {
+      // Edge case: A totally empty for loop is formatted just as `(;;)` with
+      // no splits or spaces anywhere.
+      case ForPartsWithExpression(
+            initialization: null,
+            leftSeparator: Token(precedingComments: null),
+            condition: null,
+            rightSeparator: Token(precedingComments: null),
+            updaters: NodeList(isEmpty: true),
+          )
+          when node.rightParenthesis.precedingComments == null:
+        token(forLoopParts.leftSeparator);
+        token(forLoopParts.rightSeparator);
+      case ForPartsWithDeclarations():
+        partsList.addCommentsBefore(forLoopParts.variables.beginToken);
+        visit(forLoopParts.variables);
+        finishForParts(forLoopParts, partsList);
+
+      case ForPartsWithExpression():
+        if (forLoopParts.initialization case var initializer?) {
+          partsList.addCommentsBefore(initializer.beginToken);
+          visit(initializer);
+        } else {
+          partsList.addCommentsBefore(forLoopParts.leftSeparator);
+        }
+        finishForParts(forLoopParts, partsList);
```

Comment:
[Optional] I still like flattening this one.

```suggestion
      case ForPartsWithExpression(:var initialization?):
        partsList.addCommentsBefore(initialization.beginToken);
        visit(initialization);
        finishForParts(forLoopParts, partsList);

      case ForPartsWithExpression(:var leftSeparator):
        partsList.addCommentsBefore(leftSeparator);
        finishForParts(forLoopParts, partsList);
```

Reply:
OK, I'm persuaded. Fixed locally.